Parser/src/main/java/Grammar.java

```diff
@@ -0,0 +1,107 @@
+import lombok.Getter;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Getter
+public class Grammar {
+
+    private NonTerminal initialSymbol;
+    private List<NonTerminal> nonTerminals;
+    private List<Terminal> terminals;
+    private List<Production> productions = new ArrayList<>();
+
+    private Pair<List<String>, List<String>> splitProduction(String production) {
+        List<String> splittedProduction = List.of(production.split("->"));
+        return new Pair<>(Arrays.stream(splittedProduction.get(0).strip().split(" ")).collect(Collectors.toList()),
+                Arrays.stream(splittedProduction.get(1).strip().split(" ")).collect(Collectors.toList()));
+
+    }
+
+    public void enhance() {
+        var oldInitial = this.initialSymbol;
+        this.initialSymbol = new NonTerminal(this.initialSymbol.getLabel() + "'");
+        this.productions.add(new Production(List.of(initialSymbol), List.of(oldInitial)));
+        this.nonTerminals.add(this.initialSymbol);
+    }
+
+    public boolean isNonTerminal(String label) {
+        return this.nonTerminals.stream().map(NonTerminal::getLabel).anyMatch(l -> Objects.equals(label, l));
+    }
+
+    private void loadProductions(List<String> productions) throws ReadGrammarException {
+        String production;
+        Pair<List<String>, List<String>> splittedProduction;
+        for (var i = 0; i < productions.size(); i++) {
+            production = productions.get(i);
+            splittedProduction = splitProduction(production);
```

Comment:
In general it's good to have the methods used inside another method right below it, in the order in which they are used, such that the code can be read like a newspaper🤌🏼

Reply:
e.g. move `splitProduction` below `loadProduction`